cpp/src/sort/segmented_sort.cu

```diff
@@ -0,0 +1,233 @@
+  keys_with_segid.push_back(
+    column_view(data_type(type_to_id<size_type>()), segment_ids.size(), segment_ids.data()));
+  keys_with_segid.insert(keys_with_segid.end(), keys.begin(), keys.end());
+  auto segid_keys = table_view(keys_with_segid);
+
+  std::vector<order> child_column_order(column_order);
+  if (not column_order.empty())
+    child_column_order.insert(child_column_order.begin(), order::ASCENDING);
+  std::vector<null_order> child_null_precedence(null_precedence);
+  if (not null_precedence.empty())
+    child_null_precedence.insert(child_null_precedence.begin(), null_order::AFTER);
+
+  // return sorted order of child columns
+  return detail::sorted_order(segid_keys, child_column_order, child_null_precedence, stream, mr);
+}
+
+std::unique_ptr<table> segmented_sort(table_view const& values,
+                                      table_view const& keys,
+                                      column_view const& segment_offsets,
+                                      std::vector<order> const& column_order,
+                                      std::vector<null_order> const& null_precedence,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(values.num_rows() == keys.num_rows(),
+               "Mismatch in number of rows for values and keys");
+  auto sorted_order = segmented_sorted_order(keys,
+                                             segment_offsets,
+                                             column_order,
+                                             null_precedence,
+                                             stream,
+                                             rmm::mr::get_current_device_resource());
+
+  // Gather segmented sort of child value columns`
+  return detail::gather(values,
+                        sorted_order->view(),
+                        out_of_bounds_policy::DONT_CHECK,
+                        detail::negative_index_policy::NOT_ALLOWED,
+                        stream,
+                        mr);
+}
+
+std::unique_ptr<table> sort_lists(table_view const& values,
+                                  table_view const& keys,
+                                  std::vector<order> const& column_order,
+                                  std::vector<null_order> const& null_precedence,
+                                  rmm::cuda_stream_view stream,
+                                  rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(keys.num_columns() > 0, "keys table should have atleast one list column");
+  std::vector<column_view> key_value_columns;
+  key_value_columns.reserve(keys.num_columns() + values.num_columns());
+  key_value_columns.insert(key_value_columns.end(), keys.begin(), keys.end());
+  key_value_columns.insert(key_value_columns.end(), values.begin(), values.end());
+  validate_list_columns(table_view{key_value_columns}, stream);
```

Comment:
My first thought was ITM, but then I also noticed that `key_value_columns` doesn't get used anywhere except line 159 in `validate_list_columns`. Seems like we should have a function `valid_key_value_columns` or something like that?

Reply:
making them as table_view has 2 benefits. 
1. checks key and values columns have same number of rows. (during table_view creation)
2. use parallelism across combined key+value columns.
moved it to `valid_key_value_list_columns`. (it does not sound generic enough like `validate_list_columns`)

Reply:
updated.